panic: concurrent write to websocket connection(edgecore process crashes)

<!-- Please use this template while reporting a bug and provide as much info as possible. Thanks!-->


**What happened**: edgecore process exits abnormally

**What you expected to happen**:edgecore process keeps running until we stop it on our own initiative

**How to reproduce it (as minimally and precisely as possible)**:Restart edgecore process and then deployment two pods, wait for one day or more.

**Anything else we need to know?**:

**Environment**:
- KubeEdge version:v1.15.0-kubeedge-v1.0.0

- Hardware configuration:VM

- OS (e.g. from /etc/os-release):
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

- Kernel (e.g. `uname -a`):Linux pekphicprb37565 3.10.0-693.el7.x86_64 #1 SMP Tue Aug 22 21:09:27 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
- Others:

### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### 
**Stack  Info**
2019-08-12 01:47:10.046 +08:00 ^[[0;34mINFO^[[0m pleg/generic.go:103 GenericLifecycle: Relisting
ERROR: logging before flag.Parse: I0812 01:47:10.058985    8372 generic.go:460] pod waiting > 0, pending
2019-08-12 01:47:10.063 +08:00 ^[[0;34mINFO^[[0m metamanager/msg_processor.go:608 get a message {Header:{ID:477be1f4-13ea-42a3-8c15-b8d54a63cd8b ParentID: Timestamp:1565545630052 Sync:true} Router:{Source:edged Group:meta Operation:update Resource:default/nodestatus/b3001468-bc18-463b-a2be-c9bc885f4356} Content:{UID:38796d14-1df3-11e8-8e5a-286ed488f209 Status:{Capacity:map[cpu:{i:{value:8 scale:0} d:{Dec:<nil>} s:8 Format:DecimalSI} memory:{i:{value:16655581184 scale:0} d:{Dec:<nil>} s:15884Mi Format:BinarySI} pods:{i:{value:110 scale:0} d:{Dec:<nil>} s: Format:DecimalSI}] Allocatable:map[cpu:{i:{value:8 scale:0} d:{Dec:<nil>} s:8 Format:DecimalSI} memory:{i:{value:16550723584 scale:0} d:{Dec:<nil>} s: Format:BinarySI} pods:{i:{value:110 scale:0} d:{Dec:<nil>} s: Format:DecimalSI}] Phase:Running Conditions:[{Type:Ready Status:True LastHeartbeatTime:2019-08-12 01:47:10.052145751 +0800 CST m=+228694.761644553 LastTransitionTime:2019-08-12 01:47:10.052145751 +0800 CST m=+228694.761644553 Reason:EdgeReady Message:edge is posting ready status}] Addresses:[{Type:InternalIP Address:10.247.234.208} {Type:Hostname Address:b3001468-bc18-463b-a2be-c9bc885f4356}] DaemonEndpoints:{KubeletEndpoint:{Port:0}} NodeInfo:{MachineID: SystemUUID: BootID: KernelVersion:3.10.0-862.el7.x86_64 OSImage:CentOS Linux 7 (Core) ContainerRuntimeVersion:remote://1.13.1 KubeletVersion:v1.15.0-kubeedge-v1.0.0 KubeProxyVersion: OperatingSystem:linux Architecture:amd64} Images:[] VolumesInUse:[] VolumesAttached:[] Config:nil} ExtendResources:map[]}}
2019-08-12 01:47:10.073 +08:00 DEBUG dao/meta.go:49 Update result <nil>
2019-08-12 01:47:10.073 +08:00 ^[[0;34mINFO^[[0m client/metaclient.go:98 send sync message default/nodestatus/b3001468-bc18-463b-a2be-c9bc885f4356 successed and response: {{02649c77-11f5-4a04-9ca8-9a9aa82bac4e 477be1f4-13ea-42a3-8c15-b8d54a63cd8b 1565545630073 false} {edged meta response default/nodestatus/b3001468-bc18-463b-a2be-c9bc885f4356} OK}
panic: concurrent write to websocket connection

goroutine 214 [running]:
github.com/kubeedge/kubeedge/vendor/github.com/gorilla/websocket.(*messageWriter).flushFrame(0xc000f1cea0, 0xc000beba01, 0x0, 0x0, 0x0, 0x1974446, 0xc0003ea620)
        /home/edge/gopath/src/github.com/kubeedge/kubeedge/vendor/github.com/gorilla/websocket/conn.go:591 +0x77a
github.com/kubeedge/kubeedge/vendor/github.com/gorilla/websocket.(*messageWriter).Close(0xc000f1cea0, 0xc0000e8400, 0x3ef)
        /home/edge/gopath/src/github.com/kubeedge/kubeedge/vendor/github.com/gorilla/websocket/conn.go:709 +0x56
github.com/kubeedge/kubeedge/vendor/github.com/gorilla/websocket.(*Conn).prepWrite(0xc000360000, 0x2, 0x26f0917, 0x4987f40)
        /home/edge/gopath/src/github.com/kubeedge/kubeedge/vendor/github.com/gorilla/websocket/conn.go:459 +0x20d
github.com/kubeedge/kubeedge/vendor/github.com/gorilla/websocket.(*Conn).NextWriter(0xc000360000, 0x2, 0x1000000000003ef, 0x0, 0x10, 0xc000672c00)
        /home/edge/gopath/src/github.com/kubeedge/kubeedge/vendor/github.com/gorilla/websocket/conn.go:494 +0x39
github.com/kubeedge/kubeedge/vendor/github.com/gorilla/websocket.(*Conn).WriteMessage(0xc000360000, 0x2, 0xc000672c00, 0xa, 0x10, 0x0, 0x10)
        /home/edge/gopath/src/github.com/kubeedge/kubeedge/vendor/github.com/gorilla/websocket/conn.go:755 +0x73
github.com/kubeedge/kubeedge/vendor/github.com/kubeedge/viaduct/pkg/lane.(*WSLane).Write(0xc000be4c80, 0xc000672c00, 0xa, 0x10, 0xa, 0xc000672c00, 0x0)
        /home/edge/gopath/src/github.com/kubeedge/kubeedge/vendor/github.com/kubeedge/viaduct/pkg/lane/ws.go:56 +0x5a
github.com/kubeedge/kubeedge/vendor/github.com/kubeedge/viaduct/pkg/packer.(*Writer).Write(0xc000bebd00, 0xc0000e8400, 0x3ef, 0x3ef, 0x3ef, 0x3ef, 0x0)
        /home/edge/gopath/src/github.com/kubeedge/kubeedge/vendor/github.com/kubeedge/viaduct/pkg/packer/writer.go:37 +0x172
github.com/kubeedge/kubeedge/vendor/github.com/kubeedge/viaduct/pkg/lane.(*WSLane).WriteMessage(0xc000be4c80, 0xc001218f00, 0x0, 0x0)
        /home/edge/gopath/src/github.com/kubeedge/kubeedge/vendor/github.com/kubeedge/viaduct/pkg/lane/ws.go:71 +0x10c
github.com/kubeedge/kubeedge/vendor/github.com/kubeedge/viaduct/pkg/conn.(*WSConnection).WriteMessageAsync(0xc00043dcb0, 0xc001218f00, 0x1, 0x2e87f47)
        /home/edge/gopath/src/github.com/kubeedge/kubeedge/vendor/github.com/kubeedge/viaduct/pkg/conn/ws.go:176 +0xb3
github.com/kubeedge/kubeedge/edge/pkg/edgehub/clients/wsclient.(*WebSocketClient).Send(0xc00081c160, 0xc001171e60, 0x24, 0x0, 0x0, 0x16c81ca8964, 0x1, 0x2e87f47, 0x5, 0x2e87230, ...)
        /home/edge/gopath/src/github.com/kubeedge/kubeedge/edge/pkg/edgehub/clients/wsclient/websocket.go:93 +0x7d
github.com/kubeedge/kubeedge/edge/pkg/edgehub.(*Controller).sendToCloud(0xc000304460, 0xc001171e60, 0x24, 0x0, 0x0, 0x16c81ca8964, 0x1, 0x2e87f47, 0x5, 0x2e87230, ...)
        /home/edge/gopath/src/github.com/kubeedge/kubeedge/edge/pkg/edgehub/controller.go:189 +0x69
github.com/kubeedge/kubeedge/edge/pkg/edgehub.(*Controller).routeToCloud(0xc000304460)
        /home/edge/gopath/src/github.com/kubeedge/kubeedge/edge/pkg/edgehub/controller.go:226 +0x17f
created by github.com/kubeedge/kubeedge/edge/pkg/edgehub.(*Controller).Start
        /home/edge/gopath/src/github.com/kubeedge/kubeedge/edge/pkg/edgehub/controller.go:86 +0x17e

--- a/edge/pkg/edgehub/controller.go
+++ b/edge/pkg/edgehub/controller.go
@@ -188,7 +188,9 @@ func (ehc *Controller) routeToEdge() {
 }
 
 func (ehc *Controller) sendToCloud(message model.Message) error {
+	ehc.keeperLock.Lock()
 	err := ehc.chClient.Send(message)
+	ehc.keeperLock.Unlock()
 	if err != nil {
 		klog.Errorf("failed to send message: %v", err)
 		return fmt.Errorf("failed to send message, error: %v", err)
@@ -239,12 +241,15 @@ func (ehc *Controller) keepalive() {
 		msg := model.NewMessage("").
 			BuildRouter(ModuleNameEdgeHub, "resource", "node", "keepalive").
 			FillBody("ping")
-		err := ehc.chClient.Send(*msg)
+
+		// post message to cloud hub
+		err := ehc.sendToCloud(*msg)
 		if err != nil {
 			klog.Errorf("websocket write error: %v", err)
 			ehc.stopChan <- struct{}{}
 			return
 		}
+
 		time.Sleep(ehc.config.HeartbeatPeriod)
 	}
 }

